### PR TITLE
Propose pattern to inject enterprise jobs into search pipeline

### PR DESCRIPTION
**Pull request by eseliger**
With Own, we have a need to inject enterprise code into the search pipeline. After a few different attempts, I have settled on this approach. It does not feel _too_ invasive (well of course it does but not much more than anything else in our annoying enterprise/oss split, you know) and is fairly easy to adopt in other services outside of frontend, too which is showcased in the worker job for codemonitors, for example. I've made this non-own specific so in the future we can hook in more jobs here. I've also implemented an unimplemented version (lol that sounds weird) of it so that users properly get informed when a search job type is not supported on the instance. Every job creator can decide to return a NoopJob instead, but for Own we find it more natural than instead trying to find files called `has.owner(erik)`.

Let me know what you think about this pattern! If this flies well, I'll be able to move all the backend own code into enterprise after this lands :)

## Test plan

Verified search still works, that own still works, and that own stops working in OSS.

### CODE REVIEW

**Review by camdencheek — APPROVED** [review]
Wow, this is painful! This is a perfect poster child PR for the pain caused by the enterprise split. And I really think this is a good example of taking the disciplined approach and not just adding another noodle to our big plate of globally-mutated init-time spaghetti. The changes look good -- thanks for taking the time to make a clean move. Please don't let my remaining comments block merging.

All that said, I think this is a good opportunity to brainstorm alternatives. [I don't actually like this approach](https://sourcegraph.slack.com/archives/C0379USFD7E/p1677261778468059?thread_ts=1677260340.653839&cid=C0379USFD7E), I just think it's the best we've got. Having to thread enterprise state all the way through the backend is _painful_, and in most cases, we just _don't_ and resort to overriding globals like I mentioned above. 

To start the conversation, I'd like to propose a not-fully-thought-out alternate implementation of this PR in a world where we don't have to put all enterprise code in the `enterprise` subdirectory.

Basically, rather than threading the set of enterprise implementations through the whole app, we could instead gate enterprise features with a build flag and provide alternative implementations local to the packages they are relevant to. 

In this case, we would add two new files to the `internal/search/job/jobutil` package: `owners.go` and `owners_enterprise.go`.

The `owners.go` implementation would have the default, non-enterprise implementation of the owner-related jobs. In this case, `FileHasOwnerJob` and `SelectFileOwnerJob`. This would be gated behind a build tag with `//go:build !enterprise`.

The `owners_enterprise.go` implementation would have the enterprise-licensed code. It would be gated behind the `//go:build enterprise` build tag and would include either a copy of the license or a notice at the top that this code is enterprise licensed (I don't know if this is necessary, but this seems to be common practice for split-license codebases). This file can import from other enterprise-licensed packages and use things in other files that are protected by the `//go:build enterprise` build flag.

Notice that this doesn't require _any_ threading of enterprise alternatives through the codebase. Additionally, the isolation happens at build time, not init time. You could use the same `main` for both OSS and enterprise, and switching to an enterprise build is as easy as `go build -tags enterprise`.

Anyways...I'll probably write an RFC for this post-Starship. Just wanted to get the conversation started.

**Comment by eseliger** [comment]
That all sounds very reasonable to me, thanks for sharing your idea here! Very supportive of trying _something_ else here, too. 
I'll go ahead with this one for now and am hoping we can rip it out gain somewhat soon-ish!

File: enterprise/internal/search/jobs.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+func (e *enterpriseJobs) FileHasOwnerJob(child job.Job, includeOwners, excludeOwners []string) job.Job {
+	return search.NewFileHasOwnersJob(child, includeOwners, excludeOwners)
+}
+
+func (e *enterpriseJobs) SetSelectFileOwnerJob(j func(child job.Job) job.Job) {
+}
+
+func (e *enterpriseJobs) SelectFileOwnerJob(child job.Job) job.Job {
+	return search.NewSelectOwnersJob(child)
+}
```
Comment (bobheadxi) [inline enterprise/internal/search/jobs.go:27]: This is the core of the approach, right? Thinking of its implications for https://github.com/sourcegraph/sourcegraph/pull/47493

Comment (eseliger) [reply enterprise/internal/search/jobs.go:27]: yep! Forgot to remove the unnecessary setters though, PR here: https://github.com/sourcegraph/sourcegraph/pull/48494
You can add any new enterprise Job and register it in here, and utilize it on the OSS end in search.

**Pull request merged**